### DISCUSSION TITLE
Allow overriding user language primary keys used

### DISCUSSION THREAD

**Discussion by iksent** [directus/directus#12792] [Feature Requests]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

See steps to reproduce

### To Reproduce

1. Create a model with auto-configured translations
2. Set "Use Current User Language" to "Enable" at "Translations" at "Display" tab:
![image](https://user-images.githubusercontent.com/32956521/153243933-77eb7975-4adf-4266-a03e-507514dda7dd.png)
3. Create 2 translations:
![image](https://user-images.githubusercontent.com/32956521/153244221-85b2e224-149c-496d-8f7a-2eb3e4faffba.png)
4. Change profile language to something else from English (I choose Russian):
![image](https://user-images.githubusercontent.com/32956521/153244305-2ec25d63-31c5-4e6d-b98a-8161ffca2bd7.png)
5. Go to model items table and see, that there is an "English" default item displayed instead of Russian:
![image](https://user-images.githubusercontent.com/32956521/153244453-89af49bf-8eb2-4ab4-b298-ab3cbc3f0c97.png)

### What version of Directus are you using?

9.5.1

### What version of Node.js are you using?

14.17.5

### What database are you using?

MariaDB

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying Directus?

running locally

**Comment by azrikahar** [comment]
I can't seem to reproduce this neither on `v9.5.0` nor the latest version `v9.5.1` (I'd still recommend actually confirming this on the latest version in accordance to the **Preflight Checklist** in this and future issues 👍):

https://user-images.githubusercontent.com/42867097/153353403-f5b907bc-16e0-4c08-97d5-4a3dc552873b.mp4

Not sure what could be happening here, but I'd also recommend trying to replicate this on a fresh Directus instance + fresh DB on your end and see whether it happens or not.

**Comment by iksent** [comment]
Sorry, we are on the latest 9.5.1, will change it in the description now.

Weird! I just moved my DB to the remote server and the issue is still persists...

Should I send you my database?

**Comment by iksent** [comment]
Oh! I just found that it displays chaotically:

![image](https://user-images.githubusercontent.com/32956521/153355934-46564865-2fdc-4d20-9ce9-b000029ab62b.png)

Can you please try to create more items?

**Comment by azrikahar** [comment]
> Sorry, we are on the latest 9.5.1, will change it in the description now.

> Weird! I just moved my DB to the remote server and the issue is still persists...

No worries! Would it be possible for you to test on a fresh instance as a sanity check? But feel free to hold off on this first as I think we should approach this from the following steps first.

> Should I send you my database?

That is definitely a great plan, but I do wish to try the following checks first since honestly the possibilities of things _outside_ of user or userStore in Pinia affecting this is pretty unlikely.

Currently the language display grabs the user's language, so it'll be helpful if you can verify the `/users/me` endpoint upon refresh shows the correct language before/after language switches:

https://user-images.githubusercontent.com/42867097/153358431-cc852e75-045a-44bd-baba-311bcc8c7575.mp4


> Oh! I just found that it displays chaotically:

> Can you please try to create more items?

Tried to create more items as shown above 👍 Still wondering if this is something purely app side or relates to the returned API result. Perhaps you can try with another browser as well just to rule out any chance of browser-related issues.

**Comment by iksent** [comment]
1) Yes, `/me` endpoint returns correct language: 
<img src="https://user-images.githubusercontent.com/32956521/153360013-2ae600e0-5e1d-43df-adc3-6fefc592f0b3.png" width="500px"/>

2) Tried Safari (instead of Chrome), same problem:

<img src="https://user-images.githubusercontent.com/32956521/153360202-d9e479fb-1413-40d5-83c7-68a1473f5aab.png" width="400px"/>

**Comment by iksent** [comment]
I've just figured it out, that `sort` param is set to ID:
![image](https://user-images.githubusercontent.com/32956521/153361418-864febff-35b6-4974-8e93-5d88e5261bc9.png)

And this probably leads to incorrect display:

![image](https://user-images.githubusercontent.com/32956521/153361528-e0c7b96b-69f8-47b4-9410-d543a4486f16.png)

So try to create RU/DE items firstly (start to filling them first)?

**Comment by azrikahar** [comment]
I can confirm the _returned API_ does indicate that they are in order of creation, but the Translations display seems to still be working as intended in my case:

https://user-images.githubusercontent.com/42867097/153374514-244b884a-6be5-4a20-a359-822a9a764acc.mp4

---

Took a second look and it seems like your languages table uses `languages_code` rather than `languages_id` (auto generated):

![image](https://user-images.githubusercontent.com/42867097/153377965-581a87ca-94f2-4384-a2ec-3ef26473fb92.png)

My current running theory would be _somehow_ there's something funky with your translation/language relation thus it is failing the logic here:

https://github.com/directus/directus/blob/f4275b41b36c747e34125d134baeffd6c08df837/app/src/displays/translations/translations.vue#L103

and since the Left Hand Side with the `langPkField` fails, your current display now defaults to the RHS which is the first translation item.

Would you mind taking a schema snapshot and share it here? Feel free to use the following template since GH doesn't allow `.yaml` file uploads (and I personally feel although zip file works, it's also nice to be able to see the schema at a glance without needing to download + unzip it 😁)

``````
<details>
<summary>Schema Snapshot</summary>

```yaml
# paste your yaml here
```

</details>
``````

**Comment by iksent** [comment]
@azrikahar sorry i think my db is private. Can I email you personally?

**Comment by azrikahar** [comment]
No worries! I think we may not need to anymore as I just realized something... I believe what's happening here is because your language codes are `en` & `ru`, but your userStore (as we verified before) are returning `en-US` & `ru-RU`. Would you be able to verify this, with backup of course, and change your language codes according to the `xx-XX` format?

**Comment by iksent** [comment]
Yes, we changed it specially to "en" and "ru" to match our client's locales at Next.js, which is "en" & "ru" by default (and also we don't need to separate en-us from en-gb)

I will try to do it asap and back with a news ;)

**Comment by iksent** [comment]
@azrikahar, you were right!
This was the problem: changing languages codes to `en-US` & `ru-RU` solves the problem.

How can we fix this problem? Is it will be correct to try match languages based on lang prefix "en-" & "ru-"? It will solve my problem, but this is not a 100% universal solution (if such a solution exists).

-----

Another interesting issue is that "Code" field editing is allowed in my configs:

![image](https://user-images.githubusercontent.com/32956521/153698977-662df44b-e9f6-4a96-b2dd-ee4e3d55b48f.png)

But I can't edit it in the UI:

![image](https://user-images.githubusercontent.com/32956521/153698991-ca4eb311-4961-41b1-a0c2-cf099e147f28.png)

**Comment by azrikahar** [comment]
Thanks for confirming!

> Is it will be correct to try match languages based on lang prefix "en-" & "ru-"?

This does sound like a plausible solution. There is sort of a similar logic applied when we're matching locales from date-fns and fullcalendar:

https://github.com/directus/directus/blob/0537663ab2d3df7b655e5dbe03794b9c83c56881/app/src/utils/get-date-fns-locale/get-date-fns-locale.ts#L9

https://github.com/directus/directus/blob/0537663ab2d3df7b655e5dbe03794b9c83c56881/app/src/utils/get-fullcalendar-locale/get-fullcalendar-locale.ts#L6

>  It will solve my problem, but this is not a 100% universal solution (if such a solution exists).

Technically ya it might be truly universal if we can allow "mappings", such as "abc" equals "en-US", "def" equals "ru-RU", but I think that'd end up being (potentially) confusing & likely doesn't conform to the [80/20 rule](https://docs.directus.io/contributing/introduction/#feature-requests).

> But I can't edit it in the UI:

It's currently it is not possible to edit primary keys (although IIRC it was briefly discussed before on making it possible).

---

Alternatively, we can make the display item based on "language indicator field OR primary key" by modifying the logic here:

https://github.com/directus/directus/blob/0537663ab2d3df7b655e5dbe03794b9c83c56881/app/src/displays/translations/translations.vue#L93

so that for your use case, users can add an additional field to indicate `en` is mapped to `en-US`, and you'll use the `en-US` field as the displayed field.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
